kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/NamespaceBundleOwnershipListenerImpl.java

```diff
@@ -21,96 +21,164 @@
 import lombok.extern.slf4j.Slf4j;
 import org.apache.pulsar.broker.namespace.NamespaceBundleOwnershipListener;
 import org.apache.pulsar.broker.namespace.NamespaceService;
+import org.apache.pulsar.broker.service.BrokerService;
+import org.apache.pulsar.broker.service.TopicEventsListener;
 import org.apache.pulsar.common.naming.NamespaceBundle;
 import org.apache.pulsar.common.naming.NamespaceName;
 import org.apache.pulsar.common.naming.TopicDomain;
 import org.apache.pulsar.common.naming.TopicName;
 
 @AllArgsConstructor
 @Slf4j
-public class NamespaceBundleOwnershipListenerImpl implements NamespaceBundleOwnershipListener {
+public class NamespaceBundleOwnershipListenerImpl {
+
+    private static final boolean USE_TOPIC_EVENT_LISTENER = true;
 
     private final List<TopicOwnershipListener> topicOwnershipListeners = new CopyOnWriteArrayList<>();
     private final NamespaceService namespaceService;
+    private final BrokerService brokerService;
     private final String brokerUrl;
 
+    private volatile boolean closed;
+
+    private final InnerNamespaceBundleOwnershipListener bundleBasedImpl = new InnerNamespaceBundleOwnershipListener();
+
+    public NamespaceBundleOwnershipListenerImpl(BrokerService brokerService) {
+        this.brokerService = brokerService;
+        this.brokerUrl =
+                brokerService.pulsar().getBrokerServiceUrl();
+        this.namespaceService = brokerService.pulsar().getNamespaceService();
+    }
+
     /**
      * @implNote Like {@link NamespaceService#addNamespaceBundleOwnershipListener}, when a new listener is added, the
      * `onLoad` method should be called on each owned bundle if `test(bundle)` returns true.
      */
     public void addTopicOwnershipListener(final TopicOwnershipListener listener) {
         topicOwnershipListeners.add(listener);
-        namespaceService.getOwnedServiceUnits().stream().filter(this).forEach(this::onLoad);
+        namespaceService.getOwnedServiceUnits()
+                .stream()
+                .filter(bundleBasedImpl).forEach(bundleBasedImpl::onLoad);
     }
 
-    @Override
-    public void onLoad(NamespaceBundle bundle) {
-        log.info("[{}] Load bundle: {}", brokerUrl, bundle);
-        getOwnedPersistentTopicList(bundle).thenAccept(topics -> {
-            topicOwnershipListeners.forEach(listener -> {
-                if (!listener.test(bundle.getNamespaceObject())) {
-                    return;
+    private class InnerTopicEventListener implements TopicEventsListener {
+
+        @Override
+        public void handleEvent(String topicName, TopicEvent event, EventStage stage, Throwable t) {
+            log.debug("handleEvent {} {} on {}", event, stage, topicName, new Exception().fillInStackTrace());
+            if (closed) {
+                return;
+            }
+            if (stage == EventStage.SUCCESS) {
+                TopicName topicName1 = TopicName.get(topicName);
+                switch (event) {
+                    case UNLOAD:
```

Comment:
this is the tricky part.
we still have to do it the "old way" because LOAD is not triggered when the bundle is assigned to the broker.

Currently when the bundle is loaded we eagerly find all the topics in the bundle and we trigger the initialisation.
We need to do it because the GroupCoordinator and the TransactionCoordinator bootstrap when the partitions are loaded.

Tomorrow I will try to improve the loading path by loading only the "system topics" (from S4K point of view) and not all the topics in the bundle.

Reply:
as I understand, with current implementation, at least for the non-system topics:

1. bundle assigned - handled as topic load
2. admin.topics().unload - topic is unloaded, unload event is handled, but topic is not unassigned from the bundle
3. producer tries producing - topic is loaded, topic load event is missed (no bundle assignment events)

Reply:
> 1. bundle assigned - handled as topic load

No. Bundle assigned. Nothing happens. but S4K force loading all the topics in the bundle. So eventually the LOAD event is triggered.
We are going to ignore the LOAD events from the TopicEventListeners, because it is a duplicate.

> admin.topics().unload - topic is unloaded, unload event is handled, but topic is not unassigned from the bundle

Correct. Here the TopicEventListener saves the date. We are now able to detect this type of events.

> producer tries producing - topic is loaded, topic load event is missed (no bundle assignment events)

This is not accurate. When you try producing the topic is loaded and we get the LOAD event.
But in S4K we are interested in LOAD events only in order to bootstrap the "Coordinators", here:
- https://github.com/datastax/starlight-for-kafka/blob/9bf9c17338a880ad09107b4838a25a149c6f6d2e/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaProtocolHandler.java#L427
- https://github.com/datastax/starlight-for-kafka/blob/9bf9c17338a880ad09107b4838a25a149c6f6d2e/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaProtocolHandler.java#L368